mmv1/templates/cai2hcl/resource_converter.go.erb

```diff
@@ -0,0 +1,130 @@
+
+<%- # "goimports" is used to remove unreferenced imports. -%>
+import (
+    "fmt"
+    "log"
+    "reflect"
+    "time"
+    apiComputeV1 "google.golang.org/api/compute/v1"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+    "github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
+    tpg "github.com/hashicorp/terraform-provider-google-beta/google-beta"
+    "<%= import_path() -%>/tpgresource"
+    transport_tpg "<%= import_path() -%>/transport"
+    "<%= import_path() -%>/verify"
+   "github.com/GoogleCloudPlatform/terraform-google-conversion/v2/caiasset"
+   cai2hclCommon "github.com/GoogleCloudPlatform/terraform-google-conversion/v2/cai2hcl/common"
+)
+
+<%
+    resource_name = product_ns + object.name
+    
+	properties = object.all_user_properties
+
+    product_backend_name=@base_url.split('://')[1].split('.googleapis.com')[0]
+
+    converter_name = resource_name + 'Converter'
+
+	api_type = "apiComputeV1." + object.name
+%>
+
+<%=lines(compile(pwd + '/' + object.custom_code.constants)) if object.custom_code.constants-%>
+
+const <%=resource_name-%>AssetType string = "<%=product_backend_name.downcase-%>.googleapis.com/<%=object.name-%>"
+
+type <%=converter_name-%> struct {
+	name   string
+	schema map[string]*schema.Schema
+}
+
+func New<%=converter_name-%>(name string) cai2hclCommon.Converter {
+	schema := tpg.Provider().ResourcesMap[name].Schema
+
+	return &<%=converter_name-%>{
+		name:   name,
+		schema: schema,
+	}
+}
+
+func (c *<%=converter_name-%>) Convert(assets []*caiasset.Asset) ([]*cai2hclCommon.HCLResourceBlock, error) {
+	var blocks []*cai2hclCommon.HCLResourceBlock
+	config := cai2hclCommon.NewConfig()
+
+	for _, asset := range assets {
+		if asset == nil {
+			continue
+		}
+		if asset.Resource != nil && asset.Resource.Data != nil {
+			block, err := c.convertResourceData(asset, config)
+			if err != nil {
+				return nil, err
+			}
+			blocks = append(blocks, block)
+		}
+	}
+	return blocks, nil
+}
+
+func (c *<%=converter_name-%>) convertResourceData(asset *caiasset.Asset, config *transport_tpg.Config) (*cai2hclCommon.HCLResourceBlock, error) {
+	if asset == nil || asset.Resource == nil || asset.Resource.Data == nil {
+		return nil, fmt.Errorf("asset resource data is nil")
+	}
+
+	assetResourceData := asset.Resource.Data
+	var resource *<%=api_type-%>
+
+	if err := cai2hclCommon.DecodeJSON(assetResourceData, &resource); err != nil {
+		return nil, err
+	}
+	
+    hcl, _ := resource<%=resource_name-%>Read(assetResourceData, config)
```

Comment:
this should be based on `resource` rather than `asset.Resource.Data`

Reply:
mmmm never mind, you're right, this should match how flatteners operate in https://github.com/hashicorp/terraform-provider-google/blob/eb35a5ac3572edce468ffcb05bb9313e71d75cae/google/services/compute/resource_compute_forwarding_rule.go#L734 which is on a response object, which is a `map[string]interface{}`